File: packages/sdk/src/discovery.ts
import type { FetchLike } from "./types";

export interface DiscoveryDoc {
  jwks_url: string;
  endpoints: Record<string, string>;
  policy?: { enforce_routes?: string[]; sign_routes?: string[]; sign_embed?: boolean };
  protocol?: { odin?: string; proof_version?: string };
  raw?: any;
}

export async function fetchDiscovery(baseUrl: string, fetchLike?: FetchLike): Promise<DiscoveryDoc> {
  const f = fetchLike ?? (globalThis as any).fetch;
  if (!f) throw new Error("fetchDiscovery: no fetch available");
  const url = baseUrl.replace(/\/+$/, "") + "/.well-known/odin/discovery.json";
  const res = await f(url);
  if (!res.ok) throw new Error(`fetchDiscovery: ${res.status}`);
  const j = await res.json();
  const jwks = j.jwks_url || (j.endpoints && j.endpoints.jwks);
  if (!jwks) throw new Error("fetchDiscovery: missing jwks_url");
  return {
    jwks_url: jwks,
    endpoints: j.endpoints || {},
    policy: j.policy,
    protocol: j.protocol,
    raw: j,
  };
}

File: libs/odin_core/odin/discovery.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Dict, Optional

import httpx


DISCOVERY_PATH = "/.well-known/odin/discovery.json"


def discovery_url(base_url: str) -> str:
    base = (base_url or "").rstrip("/")
    return f"{base}{DISCOVERY_PATH}"


@dataclass
class Discovery:
    jwks_url: str
    endpoints: Dict[str, str]
    policy: Dict[str, Any]
    protocol: Dict[str, Any]
    raw: Dict[str, Any]

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "Discovery":
        # Prefer explicit jwks_url; fall back to endpoints.jwks
        jwks = data.get("jwks_url") or (data.get("endpoints", {}) or {}).get("jwks")
        if not jwks:
            raise ValueError("discovery: missing jwks_url")
        return cls(
            jwks_url=jwks,
            endpoints=data.get("endpoints", {}) or {},
            policy=data.get("policy", {}) or {},
            protocol=data.get("protocol", {}) or {},
            raw=data,
        )


def fetch_discovery(
    base_url: str,
    *,
    transport: Optional[httpx.BaseTransport] = None,
    timeout: float = 10.0,
) -> Discovery:
    url = discovery_url(base_url)
    with httpx.Client(transport=transport, timeout=timeout) as client:
        r = client.get(url)
        r.raise_for_status()
        return Discovery.from_dict(r.json())
